drivers/i2c/i2c_cdns.c

```diff
@@ -0,0 +1,1067 @@
+ * actual I2C frequency generated by the calculated dividers.
+ *
+ * Return: 0 on success, negative errno otherwise.
+ */
+static int32_t cdns_i2c_calc_divs(uint32_t *f, uint32_t input_clk,
+				  uint32_t *a, uint32_t *b)
+{
+	uint32_t fscl = *f;
+	uint32_t best_fscl = *f;
+	uint32_t actual_fscl, temp;
+	uint32_t div_a, div_b;
+	uint32_t calc_div_a = 0, calc_div_b = 0;
+	uint32_t last_error, current_error;
+	int32_t ret = 0;
+
+	/* calculate initial estimate for divisor_a and divisor_b */
+	temp = input_clk / (CDNS_I2C_CLK_DIV_FACTOR * fscl);
+
+	/* Check if the calculated value is out of range */
+	if ((temp == 0U) || (temp > (CDNS_I2C_DIVA_MAX * CDNS_I2C_DIVB_MAX))) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	/* Initialize the last error to a large value (no error yet) */
+	last_error = UINT32_MAX;
+
+	/* Iterate over possible values for divisor_a */
+	for (div_a = 0; div_a < CDNS_I2C_DIVA_MAX; div_a++) {
+		/* Calculate the corresponding divisor_b for this div_a */
+		div_b = DIV_ROUND_UP(input_clk, CDNS_I2C_CLK_DIV_FACTOR * fscl * (div_a + 1U));
+
+		/* Skip invalid values of div_b */
+		if ((div_b < 1U) || (div_b > CDNS_I2C_DIVB_MAX)) {
+			continue;
+		}
+
+		/* Adjust div_b for zero-based indexing */
+		div_b--;
+
+		/* Calculate the actual fscl based on the current divisors */
+		actual_fscl = input_clk / (CDNS_I2C_CLK_DIV_FACTOR * (div_a + 1U) * (div_b + 1U));
+
+		/* Skip if the actual fscl exceeds the target fscl */
+		if (actual_fscl > fscl) {
+			continue;
+		}
+
+		/* Calculate the error between the target fscl and the actual fscl */
+		current_error = fscl - actual_fscl;
+
+		/* Update the best divisors if a smaller error is found */
+		if (last_error > current_error) {
+			calc_div_a = div_a;
+			calc_div_b = div_b;
+			best_fscl = actual_fscl;
+			last_error = current_error;
+		}
+	}
+
+	/* Set the output values */
+	*a = calc_div_a;
+	*b = calc_div_b;
+	*f = best_fscl;
+
+out:
+	return ret;
+}
+
+/**
+ * cdns_i2c_setclk - Set the serial clock rate for the I2C device
+ * @i2c_bus: Pointer to the I2C data structure
+ * @clk_in: I2C clock input frequency in Hz
+ *
+ * This function sets the serial clock rate for the I2C device by configuring
+ * the clock divisors in the device's control register. The device must be idle
+ * (i.e., not actively transferring data) before calling this function.
+ *
+ * The clock rate is determined by the following formula:
+ *	Fscl = Fpclk / (22 * (divisor_a + 1) * (divisor_b + 1))
+ * Where:
+ *	- Fscl is the desired I2C clock rate
+ *	- Fpclk is the input clock frequency
+ *	- divisor_a and divisor_b are the calculated divisors to achieve the desired clock rate
+ *
+ * The serial clock rate cannot exceed the input clock divided by 22. Common I2C clock
+ * rates are 100 KHz and 400 KHz.
+ *
+ * Return: 0 on success, negative error code on failure
+ */
+static int32_t cdns_i2c_setclk(struct cdns_i2c_data *i2c_bus, uint32_t clk_in)
```

Comment:
According to the user guide, I would also expect some Glitch filter calculations here based on the system clock frequency to be written in to the "Glitch Filter Control Register" (GFCR)

Reply:
For example, I see you've got a 100MHz system clock frequency in your board, I would expect it to write a `5` in to the GFCR as 5*10ns = 50ns

Reply:
Oh... wait, disregard this, I wasn't thinking straight, This would only apply if it was a target, which is not implemented with this PR 😜